### PR TITLE
Handle New Patreon Tiers

### LINKED ISSUE #205: Handle new Patreon tiers
## Motivation
The Patreon tiers changed since the app was implemented. Different tiers now have access to different things. We should make sure that both the old and new tiers result in the right things being displayed in the app.

## Plan of action
Old tiers:
- ≥ $1, <$5: Patron podcasts
- ≥ $5: Meditations + Liturgies

New tiers:
- ≥$3, < $10: Patron podcasts
- ≥ $10: Meditations + Liturgies

The tiers should be distinguishable by their names and amounts in the Patreon API (where they are called `rewards`). For instance, a $1 pledge on the old "Be a Liturgist" tier should unlock Patron podcasts, but a $1 pledge with no tier should not. This is probably just a small number of special cases atop the existing pledge-amount logic, with changed thresholds.

This work should probably start with some unit tests that provide mocked Patreon API responses and assert that the functions used in PatreonScreen.js return the right values.

## Scope
The only thing the app does with regard to pledge levels is to display the list of rewards that the user has access to. All actual access control is in the backend.

### PR DESCRIPTION
**Pull request by ericbendall**
## Description
Implements: https://github.com/theliturgists/app/issues/205
* Handles both new and old patreon tiers
* Also refactors patreon selectors / reducers so that the JsonApiDataStore is used once in the reducer when the data comes in
* Also hooks up Reactotron's networking functionality 

## Motivation and Context
Closes #205.

## How Has This Been Tested?
This final version still needs to be tested. The development server seemed to be down when I was trying to run this.

Also, we should probably make unit tests for this.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/state/ducks/patreon/reducer.js [hunk 1]
```diff
@@ -26,6 +27,7 @@ import {
 const defaultState = {
   token: null,
   loading: false,
+  user: null,
```
Comment (brettdh) [inline src/state/ducks/patreon/reducer.js]: Add to the state shape docs above plz

...which I'm now realizing is out of date 🤦‍♂ (if you could also remove `token` from that doc, that'd be greeeeat)

Comment (brettdh) [reply src/state/ducks/patreon/reducer.js]: Also, should `rawPayload` be listed here (and documented above)?

File: src/state/ducks/patreon/reducer.js [hunk 2]
```diff
@@ -48,11 +50,24 @@ export default handleActions({
...
     loading: true,
   }),
-  [STORE_DETAILS]: (state, action) => ({
-    ...state,
-    details: action.payload,
-    loading: false,
-  }),
+  [STORE_DETAILS]: (state, action) => {
+    let user = null;
+    const details = new JsonApiDataStore();
```
Comment (brettdh) [inline src/state/ducks/patreon/reducer.js]: Nit: suggest calling this `store` or `data` or `datastore` instead

File: src/state/ducks/patreon/selectors.js [hunk 3]
```diff
@@ -1,60 +1,83 @@
...
   const pledges = user.pledges.filter(
     p => p.reward.campaign.url === CAMPAIGN_URL,
   );
   return pledges.length > 0 ? pledges[0] : null;
 }
 
 export function isPatron(state) {
-  return getPledge(state) !== null;
+  const pledge = getPledge(state);
+  return pledge !== null && pledge?.reward?.id !== OLD_TIER_REWARD_IDS.NON_MEMBER;
```
Comment (brettdh) [inline src/state/ducks/patreon/selectors.js:29]: I was wondering about JS support for the optional chaining, and wondering if we were getting this via transpilation, and TIL about https://facebook.github.io/react-native/docs/javascript-environment :+1:

File: src/state/ducks/patreon/selectors.js [hunk 4]
```diff
@@ -1,60 +1,83 @@
...
 
 export function firstName(state) {
-  return _.get(state.patreon, 'details.data.attributes.first_name', '');
+  return _.get(state.patreon, 'rawPayload.data.attributes.first_name', '');
 }
 
 export function fullName(state) {
-  return _.get(state.patreon, 'details.data.attributes.full_name', 'Patreon');
+  return _.get(state.patreon, 'rawPayload.data.attributes.full_name', 'Patreon');
 }
```
Comment (brettdh) [inline src/state/ducks/patreon/selectors.js]: Any reason we don't also get these from the serialized `user` data?

**Review by brettdh — APPROVED** [review]
This is great stuff, Eric; thanks a ton. I was going to suggest you add unit tests, but since I'm going to be extracting this to a common module for reuse in the backend, I can do that if you prefer.

**Comment by brettdh** [comment]
Forgot to say - I'll pull the branch and test out the reducer changes real quick. Are you still having trouble with the staging API?

File: src/state/ducks/patreon/selectors.js [hunk 5]
```diff
@@ -1,60 +1,83 @@
...
 
-  const data = new JsonApiDataStore();
-  data.sync(state.patreon.details);
-
-  const userId = state.patreon.details.data.id;
-  const user = data.find('user', userId);
+  const user = state.patreon.user;
   const pledges = user.pledges.filter(
     p => p.reward.campaign.url === CAMPAIGN_URL,
   );
```
Comment (brettdh) [inline src/state/ducks/patreon/selectors.js]: Well, damn. Turns out this doesn't work:

```
undefined is not an object (evaluating 'user.pledges.filter')
```

`user` is raw json:api data, and `user.data.relationships.pledges` is just primary documents (`type` and `id`), with no `attributes`.

I'm thinking this is reason enough to revert the commit with the reducer changes, as it's probably a pain to put all the data in the right places to make it easy to work with - and `JsonApiDataStore` seems like it makes us choose between ease of data access and serializability.

Comment (brettdh) [reply src/state/ducks/patreon/selectors.js]: I reverted that commit locally, and while it doesn't crash anymore, it also shows the wrong rewards for my patreon account (it only shows the bonus podcast, where I'm somewhere above the new meditations tier). Will investigate further.

**Review by brettdh — CHANGES_REQUESTED** [review]

File: src/state/ducks/patreon/selectors.js (hunk 5 above)
Comment (brettdh) [reply src/state/ducks/patreon/selectors.js]: Ok, I pushed a commit reverting the reducer change, and a commit fixing the issue I described (spoiler: json:api IDs are strings, and constants.js stored integers).

If you want to take another stab at refactoring the data serialization out of the selector, I'm up for that, but maybe in a separate PR? I'm pretty happy with this one as it is now.

**Comment by brettdh** [comment]
@ericbendall I wanted to double-check before merging this: did you intend to use your sony.com email address (which you did), or a personal one?

File: src/state/ducks/patreon/selectors.js (hunk 5 above)
Comment (ericbendall) [reply src/state/ducks/patreon/selectors.js]: Thanks for taking a stab at actually running it there. That sucks that serializing doesn't work the way we thought. Boo.

Thanks for reverting and fixes those!

**Comment by ericbendall** [comment]
No, but I'll figure that out. My `~/.gitconfig` probably needs to be tweaked.

**Review by brettdh — APPROVED** [review]

**Pull request merged**